libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/lifecycle/UIViewModel.kt

```diff
@@ -0,0 +1,61 @@
+package com.mapbox.navigation.dropin.lifecycle
+
+import androidx.annotation.CallSuper
+import com.mapbox.navigation.base.ExperimentalPreviewMapboxNavigationAPI
+import com.mapbox.navigation.core.MapboxNavigation
+import com.mapbox.navigation.core.lifecycle.MapboxNavigationObserver
+import com.mapbox.navigation.utils.internal.InternalJobControlFactory
+import com.mapbox.navigation.utils.internal.JobControl
+import kotlinx.coroutines.cancelChildren
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.launch
+
+/**
+ * Implement your version of a behavior. Behaviors do not reference android ui elements directly,
+ * this is because their lifecycle will survive beyond a view or activity.
+ *
+ * Behaviors have a lifecycle, contain state, and process actions. [UIComponent] will respond
+ */
+@OptIn(ExperimentalPreviewMapboxNavigationAPI::class)
+abstract class UIViewModel<State, Action>(initialState: State) : MapboxNavigationObserver {
```

Comment:
This looks more like a [Redux Reducer](https://redux.js.org/tutorials/essentials/part-1-overview-concepts#redux-application-data-flow) than a [ViewModel](https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93viewmodel).

Reply:
Yes you are definitely correct! `UIReducer`? Right now it's stolen from the existing view model.

https://github.com/mapbox/mapbox-navigation-android/blob/c8f6f3b73ec1bba3242e8b31c7dd1c0e2c801256/libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/DropInViewModel.kt#L16

Let's bring this up in the meeting tomorrow as we figure out names for the things we want. Requirements are:

- Holds state between orientation changes
- Is lifecycle aware (onAttach onDetach)